Stock.Api/Controllers/ProviderController.cs

```diff
@@ -0,0 +1,173 @@
+using System;
+using System.Net;
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using AutoMapper;
+using Microsoft.AspNetCore.Mvc;
+using Stock.Api.DTOs;
+using Stock.Api.Extensions;
+using Stock.AppService.Services;
+using Stock.Model.Entities;
+using Stock.Model.Exceptions;
+
+namespace Stock.Api.Controllers
+{
+    [Produces("application/json")]
+    [Route("api/provider")]
+    [ApiController]
+    public class ProviderController : ControllerBase
+    {
+        private ProviderService service;
+        private readonly IMapper mapper;
+
+        public ProviderController(ProviderService service, IMapper mapper)
+        {
+            this.service = service;
+            this.mapper = mapper;
+        }
+
+        /// <summary>
+        /// Permite crear una nueva instancia
+        /// </summary>
+        /// <param name="value">Una instancia</param>
+        [HttpPost]
+        public ActionResult Post([FromBody] ProviderDTO value)
+        {
+            TryValidateModel(value);
+
+            try
+            {
+                var provider = this.mapper.Map<Provider>(value);
+                this.service.Create(provider);
+                value.Id = provider.Id;
+                return Ok(new { Success = true, Message = "Created", data = value });
+            }
+            catch (NameAlreadyInUseException ex)
+            {
+                return BadRequest(new { Success = false, Message = ex.Message });
+            }
+            catch (Exception)
```

Comment:
No está mal atrapar excepciones (obvio, está bien), pero:

- ¿Por qué atraparías excepciones default (sin nombrarlas o sin tipar)?
- Al poner try-catch en cada método... ¿hay algún bad smell que estes viendo? ¿como lo arreglarías?

Reply:
Atrapar la excepción default "Exception" lo agregué para que no se me pase cualquier error general y devolver un internal server error, entiendo que sino se stoppea la app. 
Pero no me gusta nada tener tantos try catch, puede ser que tenga que usar una especie de middleware? porque no puedo tampoco estar definiendo todas las excepciones en mi controller, sino tendría que considerar excepciones de la capa de services, de data, etc.
Ref. de lo que googlie: https://briancaos.wordpress.com/2019/12/06/net-core-api-catch-exceptions-using-a-middleware/#:~:text=To%20handle%20such%20a%20scenario,obvious%20are%20as%20exception%20handlers.&text=This%20middleware%20will%20catch%20errors,as%20a%20JSON%20result%20object.

Reply:
Perfecto! hay que usar el middleware de netcore. 

Está en la carpeta Exception y la clase ExceptionMiddlware. 

La idea perfecta sería elevar una excepción "BusinessException" y atrapar esa excepción, sabiendo que no se cumplió alguna invariante de las reglas de negocio.